diff --git a/CRM/Contribute/Form/Contribution.php b/CRM/Contribute/Form/Contribution.php
index 04310298e83..fff41eba35d 100644
--- a/CRM/Contribute/Form/Contribution.php
+++ b/CRM/Contribute/Form/Contribution.php
@@ -1647,8 +1647,8 @@ protected function submit($submittedValues, $action, $pledgePaymentID) {
       }
       $params['line_item'] = $lineItem;
       $params['payment_processor_id'] = $params['payment_processor'] = CRM_Utils_Array::value('id', $this->_paymentProcessor);
-      if (isset($submittedValues['tax_amount'])) {
-        $params['tax_amount'] = $submittedValues['tax_amount'];
+      if (isset($this->_values['tax_amount'])) {
+        $params['tax_amount'] = $this->_values['tax_amount'];
       }
       //create contribution.
       if ($isQuickConfig) {
